### PR TITLE
BUG: Allow integer input data in scipy.stats.yeojohnson

### LINKED ISSUE #10421: Yeo-Johnson power transformation fails with integer input data
My issue is about [scipy.stats.yeojohnson](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.yeojohnson.html). A ValueError is raised when the input data has an integer type, because `Integers to negative integer powers are not allowed.`

### Reproducing code example:

```python
import numpy as np
from scipy.stats import yeojohnson


yeojohnson(np.arange(10))
```
which fails, while it works if the data type is np.float64:
```python
yeojohnson(np.arange(10).astype('float64))
```

### Error message:

```python
ValueError                                Traceback (most recent call last)
<ipython-input-4-2cd13f92b547> in <module>
----> 1 yeojohnson(np.arange(10))

~/anaconda3/lib/python3.7/site-packages/scipy/stats/morestats.py in yeojohnson(x, lmbda)
   1345 
   1346     # if lmbda=None, find the lmbda that maximizes the log-likelihood function.
-> 1347     lmax = yeojohnson_normmax(x)
   1348     y = _yeojohnson_transform(x, lmax)
   1349 

~/anaconda3/lib/python3.7/site-packages/scipy/stats/morestats.py in yeojohnson_normmax(x, brack)
   1518         return -yeojohnson_llf(lmbda, data)
   1519 
-> 1520     return optimize.brent(_neg_llf, brack=brack, args=(x,))
   1521 
   1522 

~/anaconda3/lib/python3.7/site-packages/scipy/optimize/optimize.py in brent(func, args, brack, tol, full_output, maxiter)
   2107     options = {'xtol': tol,
   2108                'maxiter': maxiter}
-> 2109     res = _minimize_scalar_brent(func, brack, args, **options)
   2110     if full_output:
   2111         return res['x'], res['fun'], res['nit'], res['nfev']

~/anaconda3/lib/python3.7/site-packages/scipy/optimize/optimize.py in _minimize_scalar_brent(func, brack, args, xtol, maxiter, **unknown_options)
   2139                   full_output=True, maxiter=maxiter)
   2140     brent.set_bracket(brack)
-> 2141     brent.optimize()
   2142     x, fval, nit, nfev = brent.get_result(full_output=True)
   2143     return OptimizeResult(fun=fval, x=x, nit=nit, nfev=nfev,

~/anaconda3/lib/python3.7/site-packages/scipy/optimize/optimize.py in optimize(self)
   1923         # set up for optimization
   1924         func = self.func
-> 1925         xa, xb, xc, fa, fb, fc, funcalls = self.get_bracket_info()
   1926         _mintol = self._mintol
   1927         _cg = self._cg

~/anaconda3/lib/python3.7/site-packages/scipy/optimize/optimize.py in get_bracket_info(self)
   1900         elif len(brack) == 2:
   1901             xa, xb, xc, fa, fb, fc, funcalls = bracket(func, xa=brack[0],
-> 1902                                                        xb=brack[1], args=args)
   1903         elif len(brack) == 3:
   1904             xa, xb, xc = brack

~/anaconda3/lib/python3.7/site-packages/scipy/optimize/optimize.py in bracket(func, xa, xb, args, grow_limit, maxiter)
   2322     _gold = 1.618034  # golden ratio: (1.0+sqrt(5.0))/2.0
   2323     _verysmall_num = 1e-21
-> 2324     fa = func(*(xa,) + args)
   2325     fb = func(*(xb,) + args)
   2326     if (fa < fb):                      # Switch so fa > fb

~/anaconda3/lib/python3.7/site-packages/scipy/stats/morestats.py in _neg_llf(lmbda, data)
   1516 
   1517     def _neg_llf(lmbda, data):
-> 1518         return -yeojohnson_llf(lmbda, data)
   1519 
   1520     return optimize.brent(_neg_llf, brack=brack, args=(x,))

~/anaconda3/lib/python3.7/site-packages/scipy/stats/morestats.py in yeojohnson_llf(lmb, data)
   1461         return np.nan
   1462 
-> 1463     trans = _yeojohnson_transform(data, lmb)
   1464 
   1465     loglike = -n_samples / 2 * np.log(trans.var(axis=0))

~/anaconda3/lib/python3.7/site-packages/scipy/stats/morestats.py in _yeojohnson_transform(x, lmbda)
   1362         out[pos] = np.log1p(x[pos])
   1363     else:  # lmbda != 0
-> 1364         out[pos] = (np.power(x[pos] + 1, lmbda) - 1) / lmbda
   1365 
   1366     # when x < 0

ValueError: Integers to negative integer powers are not allowed.
```

### Scipy/Numpy/Python version information:

```python
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.3.0 1.16.4 sys.version_info(major=3, minor=7, micro=1, releaselevel='final', serial=0)
```

### Proposed solution

The first operation is to convert `x` as np.array:
https://github.com/scipy/scipy/blob/e1e44d12637997606b1bcc0c6de232349e11eee0/scipy/stats/morestats.py#L1339
I think that it can be expected to have float data when doing a Yeo-Johnson power transformation, so we could just modify the dtype by adding `dtype='float64'` (or maybe find the lowest value among `float16`, `float32` and `float64`).

I would like to work on this if possible.

### PR DESCRIPTION
**Pull request by johannfaouzi**
Fixes #10421 

This small PR allows the use of integer data for [scipy.stats.yeojohnson](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.yeojohnson.html): the input data is converted into `np.float64` dtype. I am not very knowledgeable about memory use, so it might not be ideal for extremely large arrays, but I think that it can be expected to have the data converted into float type when doing this kind of transformation.

A small test is added in the YeoJohnson class.

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
Looks sensible at first glance -- if this wasn't intended behavior then there should be documentation specifying the type limitation.

I think if you rebase on the latest master branch the CI failure will go away since we removed pypy3 from our CI--it was too much of a maintenance burden.

I see that no resident stats expert has commented either on the cognate issue or here--would be good if they did briefly.

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -1329,7 +1329,7 @@ def yeojohnson(x, lmbda=None):
 
     """
 
-    x = np.asarray(x)
+    x = np.asarray(x, dtype=np.float64)
```
Comment (tylerjereddy) [inline scipy/stats/morestats.py]: As you note in the issue, some consideration for passing through smaller dtypes might be warranted. Would people typically call this function with a smaller float type and expect the return to be of the same type? The promotion logic for doing that and supporting integers might be a little messy maybe.

**Comment by johannfaouzi** [comment]
I don't really have a preference regarding the solution (converting the data to float type, possibly with a warning, versus raising an error), but the current behavior could be improved (the error message is not very clear).

Do you think that adding something like this would be better?
```python
if np.issubdtype(x.dtype, np.integer):
    raise ValueError("Yeo-Johnson transformation cannot be performed on integers. "
                     "Convert your data type to float.)
```

**Comment by johannfaouzi** [comment]
Currently the transformation works for complex numbers, which may not be the expected behavior. The following code might be more appropriate: 

```python
if not np.issubdtype(x.dtype, np.floating):
    raise ValueError("Yeo-Johnson transformation must be performed on floats. "
                     "Convert your data type to float.)
```

**Comment by rlucas7** [comment]
>I see that no resident stats expert has commented either on the cognate issue or here--would be good if they did briefly.

Not sure if I qualify as resident or a stats expert and I'll comment regardless. 

I agree with @tylerjereddy that we want to be careful with not accidentally promoting to float64 if the input was a float32 (for example). Probably want some kind of logic for that in the PR even though it *might* be messy. Overarching goal should be to maintain the types that are inputs unless they cause errors. 

On the complex number comment of @johannfaouzi 

>Currently the transformation works for complex numbers, which may not be the expected behavior.

I'm not sure if there are tests for `yeojohnson` gradients in the testing suite, do all unit tests pass if you add in the lines you suggest? 

There are routines in the unit tests which use complex numbers where applicable to get gradients. For example see the function here:

https://github.com/scipy/scipy/blob/8dba340293fe20e62e173bdf2c10ae208286692f/scipy/optimize/_numdiff.py#L204

I agree that the yeojohnson was likely intended primarily for real valued inputs but if there are cases where a gradient is needed it might be better to leave the capability in the function. Unless of course, you have found some undesirable behavior or results for complex invalued inputs.

**Comment by johannfaouzi** [comment]
The current implementation of the Yeo-Johnson log-likelihood works for real-valued and complex numbers:
https://github.com/scipy/scipy/blob/e1e44d12637997606b1bcc0c6de232349e11eee0/scipy/stats/morestats.py#L1463-L1466

The log-likelihood is always a real-valued number (it might be a complex with a null imaginary part when the input array consists of complex numbers), and the Brent optimization works. Looking briefly at the Wikipedia page, it seems to be a gradient-free optimization algorithm.

Given the implementation of the log-likelihood function, the imaginary parts have an impact on the transformation:
```python
>>> import numpy as np
>>> from scipy.stats import yeojohnson
>>> x = np.arange(10)
>>> yeojohnson(x + 1j)

(array([0.03287148+0.98935685j, 0.98844173+0.94568291j,
        1.91584736+0.91855106j, 2.82231669+0.89925149j,
        3.71267304+0.88439277j, 4.59007808+0.87236522j,
        5.45671414+0.86229017j, 6.31416326+0.85363873j,
        7.163619  +0.84606906j, 8.00601107+0.83934805j]),
 (0.9240010903270425+0j))
```

```python
>>> yeojohnson(x.astype(float))

(array([0.        , 0.89952677, 1.67649209, 2.38322961, 3.04195185,
        3.66477639, 4.25925107, 4.83048762, 5.38215489, 5.91700119]),
 0.7219640406219725)
```
I can't comment on the expected results here (and I'm definitively not a stats expert).

I am going to try to raise a ValueError if the input data type is not float and see how the tests react.

**Comment by johannfaouzi** [comment]
I don't understand why only one check failed. If someone can shed some light on this issue for me, it would be appreciated.

When the final solution will be accepted (not necessarily the current version obviously), I think that it could be applied to all the functions related to the Yeo-Johnson transformation:
* [scipy.stats.yeojohnson](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.yeojohnson.html)
* [scipy.stats.yeojohnson_llf](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.yeojohnson_llf.html)
* [scipy.stats.yeojohnson_normmax](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.yeojohnson_normmax.html)
* [scipy.stats.yeojohnson_normplot](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.yeojohnson_normplot.html)
* [scipy.stats._yeojohnson_transform](https://github.com/scipy/scipy/blob/e1e44d12637997606b1bcc0c6de232349e11eee0/scipy/stats/morestats.py#L1353-L1372)

File: scipy/stats/morestats.py [hunk 2]
```diff
@@ -1340,6 +1340,10 @@ def yeojohnson(x, lmbda=None):
     if x.size == 0:
         return x
 
+    if not np.issubdtype(x.dtype, np.floating):
```
Comment (rlucas7) [inline scipy/stats/morestats.py]: There are methods used in scipy where gradients are calculated from complex inputs. If you'd like to instead write a unit test that confirms those values that would be helpful. I don't think it is necessary to restrict to only float types here.

**Review by rlucas7 — CHANGES_REQUESTED** [review]
Left a comment about the complex number type and raising an error. I don't think we want to raise an error or exception in that case. Instead you could write a unit test that confirms the gradients from passing a complex number if you want.

**Comment by rgommers** [comment]
I don't think raising an error for integer input is correct. The transformation is well-defined for any real number, so users will expect it to work. The reason it's broken now:
```
~/Code/scipy/scipy/stats/morestats.py in _yeojohnson_transform(x, lmbda)

-> 1370         out[pos] = (np.power(x[pos] + 1, lmbda) - 1) / lmbda

ValueError: Integers to negative integer powers are not allowed.
```
is just due to an implementation detail and a missing test. 

You already have the check for if the input is an integer array. If so, just cast to `float64`. Then we can merge this. 

`complex` input on the other hand may raise an error I'd say, since we don't know what the right answer is.

**Comment by johannfaouzi** [comment]
I made the following changes:
* An error is raised if data type is complex;
* Data type is converted to `float64` if it is integer.

I added small tests for both additions.

The failing test is a crashed worker (I don't know if running it over is worth it).

**Comment by rgommers** [comment]
LGTM now, merged. Thanks @johannfaouzi

**Pull request merged**